File: algo_npu/EXEC/exec2/exec2/BB4TSP.cpp
//
// Created by 86180 on 25-10-24.
//
//
// Created by 86180 on 25-10-24.
//
#include <vector>
#include <climits>
#include <queue>
#include <algorithm>

using namespace std;

class BB4TSP {
private:
    const int NoEdge = -1;
    int minCost = INT_MAX; // 当前最小路径长度
    vector<int> bestH; // 最优路径
    long long nodeCount = 0; // 搜索节点计数
    int boundType = 1; // 界值函数类型

    struct HeapNode {
        vector<int> liveNode; // 活节点（当前路径）
        int lcost; // 下界
        int level; // 当前层级（已确定的城市数）
        int currentCost; // 当前实际路径长度

        HeapNode(const vector<int>& node, int lb, int lev, int cc)
            : liveNode(node), lcost(lb), level(lev), currentCost(cc) {}

        // 用于优先队列比较（最小堆）
        bool operator>(const HeapNode& other) const {
            return lcost > other.lcost;
        }
    };
    priority_queue<HeapNode, vector<HeapNode>, greater<HeapNode>> heap; // 优先队列存储活节点

    // 界值函数1：基于最小生成树的下界
    int computeLB1(const vector<int>& path, int level, int currentCost, const vector<vector<int>>& cMatrix, int n) {
        if (level == 1) {
            return 0;
        }

        int bound = currentCost;
        vector<bool> visited(n + 1, false);
        for (int i = 0; i < level; ++i) {
            visited[path[i]] = true;
        }

        // 计算未访问城市的最小边权和
        int minSum = 0;
        for (int i = 1; i <= n; ++i) {
            if (!visited[i]) {
                int minEdge = INT_MAX;
                for (int j = 1; j <= n; ++j) {
                    if (i != j && cMatrix[i][j] != NoEdge) {
                        minEdge = min(minEdge, cMatrix[i][j]);
                    }
                }
                if (minEdge != INT_MAX) {
                    minSum += minEdge;
                }
            }
        }

        // 添加回到起点的边的估计
        int minToStart = INT_MAX;
        for (int i = 1; i <= n; ++i) {
            if (!visited[i] && cMatrix[i][path[0]] != NoEdge) {
                minToStart = min(minToStart, cMatrix[i][path[0]]);
            }
        }

        if (minToStart != INT_MAX) {
            minSum += minToStart;
        }

        return bound + minSum;
    }

    // 界值函数2：基于每个城市的最小出边和入边
    int computeLB2(const vector<int>& path, int level, int currentCost, const vector<vector<int>>& cMatrix, int n) {
        if (level == 1) {
            return 0;
        }

        vector<bool> visited(n + 1, false);
        for (int i = 0; i < level; ++i) {
            visited[path[i]] = true;
        }

        int bound = currentCost;

        // 对于每个未访问城市，计算最小出边和入边
        for (int i = 1; i <= n; ++i) {
            if (!visited[i]) {
                int minOut = INT_MAX;
                int minIn = INT_MAX;

                for (int j = 1; j <= n; ++j) {
                    if (i != j) {
                        if (cMatrix[i][j] != NoEdge) {
                            minOut = min(minOut, cMatrix[i][j]);
                        }
                        if (cMatrix[j][i] != NoEdge) {
                            minIn = min(minIn, cMatrix[j][i]);
                        }
                    }
                }

                if (minOut != INT_MAX) {
                    bound += minOut / 2;
                }
                if (minIn != INT_MAX) {
                    bound += minIn / 2;
                }
            }
        }

        return bound;
    }

    // 界值函数3：改进的下界，考虑当前位置
    int computeLB3(const vector<int>& path, int level, int currentCost, const vector<vector<int>>& cMatrix, int n) {
        if (level == 1) {
            return 0;
        }

        vector<bool> visited(n + 1, false);
        for (int i = 0; i < level; ++i) {
            visited[path[i]] = true;
        }

        int bound = currentCost;
        int lastCity = path[level - 1];

        // 从当前城市到未访问城市的最小边
        int minFromCurrent = INT_MAX;
        for (int i = 1; i <= n; ++i) {
            if (!visited[i] && cMatrix[lastCity][i] != NoEdge) {
                minFromCurrent = min(minFromCurrent, cMatrix[lastCity][i]);
            }
        }

        if (minFromCurrent != INT_MAX) {
            bound += minFromCurrent;
        }

        // 未访问城市之间的最小边权和
        vector<int> unvisited;
        for (int i = 1; i <= n; ++i) {
            if (!visited[i]) {
                unvisited.push_back(i);
            }
        }

        for (int city : unvisited) {
            int minEdge = INT_MAX;
            for (int j = 1; j <= n; ++j) {
                if (city != j && cMatrix[city][j] != NoEdge) {
                    minEdge = min(minEdge, cMatrix[city][j]);
                }
            }
            if (minEdge != INT_MAX) {
                bound += minEdge / unvisited.size();
            }
        }

        return bound;
    }

    int computeLB(const vector<int>& path, int level, int currentCost, const vector<vector<int>>& cMatrix, int n) {
        switch (boundType) {
            case 1:
                return computeLB1(path, level, currentCost, cMatrix, n);
            case 2:
                return computeLB2(path, level, currentCost, cMatrix, n);
            case 3:
                return computeLB3(path, level, currentCost, cMatrix, n);
            default:
                return computeLB1(path, level, currentCost, cMatrix, n);
        }
    }

public:
    void setBoundType(int type) {
        boundType = type;
    }

    int bb4TSP(const vector<vector<int>>& cMatrix, int n) {
        minCost = INT_MAX;
        bestH.clear();
        nodeCount = 0;

        // 清空优先队列
        while (!heap.empty()) {
            heap.pop();
        }

        // 初始化根节点（从城市1开始）
        vector<int> liveNode;
        liveNode.push_back(1);
        int level = 1;
        int lcost = computeLB(liveNode, level, 0, cMatrix, n);
        heap.push(HeapNode(liveNode, lcost, level, 0));

        while (!heap.empty()) {
            nodeCount++;
            HeapNode current = heap.top();
            heap.pop();

            // 剪枝：如果当前下界已经大于等于已知最优解，跳过
            if (current.lcost >= minCost) {
                continue;
            }

            // 如果已经访问了所有城市
            if (current.level == n) {
                // 检查能否回到起点
                int lastCity = current.liveNode[n - 1];
                if (cMatrix[lastCity][1] != NoEdge) {
                    int totalCost = current.currentCost + cMatrix[lastCity][1];
                    if (totalCost < minCost) {
                        minCost = totalCost;
                        bestH = current.liveNode;
                    }
                }
                continue;
            }

            // 扩展子节点
            int lastCity = current.liveNode[current.level - 1];
            vector<bool> visited(n + 1, false);
            for (int city : current.liveNode) {
                visited[city] = true;
            }

            for (int nextCity = 1; nextCity <= n; ++nextCity) {
                if (!visited[nextCity] && cMatrix[lastCity][nextCity] != NoEdge) {
                    vector<int> newPath = current.liveNode;
                    newPath.push_back(nextCity);
                    int newCost = current.currentCost + cMatrix[lastCity][nextCity];
                    int newLevel = current.level + 1;
                    int newLB = computeLB(newPath, newLevel, newCost, cMatrix, n);

                    // 只有当下界小于当前最优解时才加入队列
                    if (newLB < minCost) {
                        heap.push(HeapNode(newPath, newLB, newLevel, newCost));
                    }
                }
            }
        }

        return minCost;
    }

    int getMinCost() const {
        return minCost;
    }

    vector<int> getBestPath() const {
        return bestH;
    }

    long long getNodeCount() const {
        return nodeCount;
    }
};

File: algo_npu/EXEC/exec3/AlgorithmAnalyse/src/main/java/Experiment2/BB4TSP.java
package Experiment2;

import java.util.Collections;
import java.util.LinkedList;
import java.util.Vector;

public class BB4TSP {

    int NoEdge = -1; // 表示无边
    private int minCost = Integer.MAX_VALUE; // 当前最小耗费
    public int getMinCost() { return minCost; }
    public void setMinCost(int minCost) { this.minCost = minCost; }

    private LinkedList<HeapNode> heap = new LinkedList<>(); // 最小堆（活节点表）
    private Vector<Integer> bestH = new Vector<>();         // 最优路径

    @SuppressWarnings("rawtypes")
    public static class HeapNode implements Comparable {
        Vector<Integer> liveNode; // 剩余可选城市集合
        int lcost;                // 当前路径下界
        int level;                // 已确定城市数（从1开始）
        int ccost;                // 当前实际路径耗费（到 level 为止）

        public HeapNode(Vector<Integer> node, int lb, int lev, int cost) {
            this.liveNode = new Vector<>(node);
            this.lcost = lb;
            this.level = lev;
            this.ccost = cost;
        }

        @Override
        public int compareTo(Object x) {
            int xu = ((HeapNode) x).lcost;
            if (lcost < xu) return -1;
            if (lcost == xu) return 0;
            return 1;
        }

        public boolean equals(Object x) {
            return lcost == ((HeapNode) x).lcost;
        }
    }

    /**
     * 计算当前状态的下界（紧下界）
     * 方法：每行最小出边 + 每列最小入边（排除已用边）
     */
    public int computeLB(Vector<Integer> liveNode, int level, int[][] cMatrix) {
        int n = cMatrix.length - 1;
        int lb = 0;
        boolean[] rowUsed = new boolean[n + 1];   // 行是否已使用（出发城市）
        boolean[] colUsed = new boolean[n + 1];   // 列是否已使用（到达城市）

        // 标记已确定路径中的边
        for (int i = 1; i < level; i++) {
            int from = bestH.get(i - 1);
            int to = bestH.get(i);
            rowUsed[from] = true;
            colUsed[to] = true;
        }

        // 1. 每行（出发城市）取最小出边（未使用列）
        for (int i = 1; i <= n; i++) {
            if (rowUsed[i]) continue; // 已出发过
            int minOut = Integer.MAX_VALUE;
            for (int j = 1; j <= n; j++) {
                if (!colUsed[j] && cMatrix[i][j] != NoEdge) {
                    minOut = Math.min(minOut, cMatrix[i][j]);
                }
            }
            if (minOut == Integer.MAX_VALUE) return Integer.MAX_VALUE; // 无解
            lb += minOut;
        }

        // 2. 每列（到达城市）取最小入边（未使用行）
        for (int j = 1; j <= n; j++) {
            if (colUsed[j]) continue; // 已到达过
            int minIn = Integer.MAX_VALUE;
            for (int i = 1; i <= n; i++) {
                if (!rowUsed[i] && cMatrix[i][j] != NoEdge) {
                    minIn = Math.min(minIn, cMatrix[i][j]);
                }
            }
            if (minIn == Integer.MAX_VALUE) return Integer.MAX_VALUE;
            lb += minIn;
        }

        return lb;
    }

    /**
     * 分支限界求解 TSP
     */
    public int bb4TSP(int[][] cMatrix, int n) {
        // 清空状态
        heap.clear();
        bestH.clear();
        minCost = Integer.MAX_VALUE;

        // 起始节点：所有城市可选，level=1（将要选第一个城市）
        Vector<Integer> initNode = new Vector<>();
        for (int i = 1; i <= n; i++) initNode.add(i);

        // 固定起点为 1
        bestH.add(1); // 路径从 1 开始
        Vector<Integer> childNode = new Vector<>(initNode);
        childNode.remove((Integer)1);

        int initLB = computeLB(childNode, 1, cMatrix);
        heap.add(new HeapNode(childNode, initLB, 1, 0));

        while (!heap.isEmpty()) {
            HeapNode node = heap.pollFirst(); // 取出下界最小的活节点
            if (node.lcost >= minCost) continue; // 下界已不优

            int level = node.level;
            Vector<Integer> liveNode = node.liveNode;
            int ccost = node.ccost;

            if (level == n) {
                // 最后一个城市 → 回到起点 1
                int last = bestH.get(level - 1);
                if (cMatrix[last][1] != NoEdge) {
                    int total = ccost + cMatrix[last][1];
                    if (total < minCost) {
                        minCost = total;
                        bestH.add(1); // 闭合回路
                    }
                }
                continue;
            }

            // 扩展当前节点：尝试每个剩余城市作为下一个
            for (int i = 0; i < liveNode.size(); i++) {
                int nextCity = liveNode.get(i);
                int from = bestH.get(level - 1);
                if (cMatrix[from][nextCity] == NoEdge) continue;

                // 生成子节点
                Vector<Integer> newLive = new Vector<>(liveNode);
                newLive.remove(i);
                bestH.add(nextCity);
                int newCost = ccost + cMatrix[from][nextCity];
                int newLB = newCost + computeLB(newLive, level + 1, cMatrix);
                bestH.remove(bestH.size() - 1); // 回溯

                if (newLB < minCost) {
                    heap.add(new HeapNode(newLive, newLB, level + 1, newCost));
                }
            }

            // 维护最小堆：重新排序（LinkedList 模拟优先队列）
            Collections.sort(heap);
        }

        return minCost;
    }

    /* ------------------- 测试代码 ------------------- */
    public static void main(String[] args) {
        int[][] graph = {
                {0, 0, 0, 0, 0},
                {0, 0, 10, 15, 20},
                {0, 10, 0, 35, 25},
                {0, 15, 35, 0, 30},
                {0, 20, 25, 30, 0}
        };

        BB4TSP solver = new BB4TSP();
        int minCost = solver.bb4TSP(graph, 4);
        System.out.println("最小路径长度 = " + minCost);

        // 输出路径（bestH 已包含完整回路）
        System.out.print("路径：");
        for (int i = 0; i < solver.bestH.size(); i++) {
            System.out.print(solver.bestH.get(i));
            if (i < solver.bestH.size() - 1) System.out.print(" -> ");
        }
        System.out.println();
    }
}